internal/clients/upbound/upbound.go

```diff
@@ -0,0 +1,86 @@
+package upbound
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/upbound/up-sdk-go"
+	"github.com/upbound/up-sdk-go/service/auth"
+	"golang.org/x/oauth2"
+	"k8s.io/client-go/rest"
+)
+
+// WrapRESTConfig configures the supplied REST config to use OAuth2 bearer
+// tokens fetched using the supplied Upbound session/robot token.
+func WrapRESTConfig(_ context.Context, rc *rest.Config, token []byte, _ ...string) error { // nolint:gocyclo // mostly error handling
+	ex := rc.ExecProvider
+	if ex == nil {
+		return errors.New("an identity configuration was specified but the provided kubeconfig does not have execProvider section")
+	}
+	if ex.APIVersion != "client.authentication.k8s.io/v1" {
+		return errors.New("execProvider APIVersion is not client.authentication.k8s.io/v1")
+	}
+	if ex.Command != "up" || len(ex.Args) < 2 || ex.Args[0] != "organization" || ex.Args[1] != "token" {
+		return errors.New("execProvider command is not up organization token")
+	}
+
+	// Read the org name, it could either be the 3rd argument and provided with `ORGANIZATION` env var.
+	org := ""
+	if len(ex.Args) > 2 {
+		org = ex.Args[2]
+	}
+	for _, env := range ex.Env {
+		if env.Name == "ORGANIZATION" {
+			// Env var takes precedence over the 3rd argument if both are provided.
+			org = env.Value
+			break
+		}
+	}
+	if org == "" {
+		return errors.New("organization name not provided in execProvider args or ORGANIZATION env var")
+	}
+
+	rc.ExecProvider = nil
+
+	var t *oauth2.Token
+	// DefaultTokenSource retrieves a token source from an injected identity.
+	us := &upboundTokenSource{
+		client: auth.NewClient(&up.Config{
+			Client: up.NewClient(func(client *up.HTTPClient) {
+				// TODO: This should probably be configurable.
+				client.BaseURL.Host = "auth.upbound.io"
+			}),
+		}),
+		org:          org,
+		refreshToken: string(token),
+	}
+	ts := oauth2.ReuseTokenSource(t, us)
```

Comment:
This is executed in each `Connect,` and we initialize a reuse token source with an initial token of `nil`, which refreshes/exchanges tokens before they actually expire (in each reconcile instead if once in an hour).  It looks like the issue already exists with other identity methods as well, e.g. Google and Azure, so we should fix them for all.

Reply:
This is resolved by introducing a `ReuseTokenSource Store` initialized during Setup.